File: rust/src/cli_args.rs
use clap::Parser;
use log::info;

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
pub struct CommandLineArgs {
    #[arg(
        long,
        value_delimiter = ',',
        help = "Comma-separated list of Reddit post URLs"
    )]
    pub urls: Vec<String>,

    #[arg(
        long = "src-files",
        value_delimiter = ',',
        help = "Comma-separated list of file paths containing URLs"
    )]
    pub src_files: Vec<String>,

    #[arg(
        long,
        value_delimiter = ',',
        help = "Comma-separated list of subreddits (e.g., r/python,r/askreddit)"
    )]
    pub subs: Vec<String>,

    #[arg(
        long,
        value_delimiter = ',',
        help = "Comma-separated list of multireddits (e.g., m/programming)"
    )]
    pub multis: Vec<String>,
}

impl CommandLineArgs {
    pub fn parse_args() -> Self {
        let args = CommandLineArgs::parse();

        info!("Parsed {} URL(s) from --urls", args.urls.len());
        info!("Parsed {} file(s) from --src-files", args.src_files.len());
        info!("Parsed {} subreddit(s) from --subs", args.subs.len());
        info!("Parsed {} multireddit(s) from --multis", args.multis.len());

        args
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_command_line_args_default() {
        let args = CommandLineArgs {
            urls: vec![],
            src_files: vec![],
            subs: vec![],
            multis: vec![],
        };

        assert_eq!(args.urls.len(), 0);
        assert_eq!(args.src_files.len(), 0);
        assert_eq!(args.subs.len(), 0);
        assert_eq!(args.multis.len(), 0);
    }

    #[test]
    fn test_command_line_args_with_data() {
        let args = CommandLineArgs {
            urls: vec!["https://www.reddit.com/r/rust/comments/test/".to_string()],
            src_files: vec!["/tmp/urls.txt".to_string()],
            subs: vec!["r/rust".to_string(), "r/programming".to_string()],
            multis: vec!["m/programming".to_string()],
        };

        assert_eq!(args.urls.len(), 1);
        assert_eq!(args.src_files.len(), 1);
        assert_eq!(args.subs.len(), 2);
        assert_eq!(args.multis.len(), 1);

        assert_eq!(args.urls[0], "https://www.reddit.com/r/rust/comments/test/");
        assert_eq!(args.subs[0], "r/rust");
        assert_eq!(args.subs[1], "r/programming");
    }
}

File: python/cli_args.py
import argparse
import logging
from typing import List
from colored_logger import get_colored_logger

logger = get_colored_logger(__name__)


def _parse_csv(csv_str: str) -> List[str]:
    """
    Splits a comma-separated string into a list of non-empty items, stripping whitespace.
    """
    if not csv_str:
        return []
    return [item.strip() for item in csv_str.split(",") if item.strip()]


class CommandLineArgs:
    """
    Parses command-line arguments in a manner similar to the original Ruby script.
    """

    def __init__(self) -> None:
        self.parser = argparse.ArgumentParser(
            description="Save the content (body and replies) of a Reddit post to Markdown or HTML."
        )
        self.parser.add_argument(
            "--urls",
            type=str,
            default="",
            help="Comma-separated list of Reddit post URLs.",
        )
        self.parser.add_argument(
            "--src-files",
            type=str,
            default="",
            help="Comma-separated list of file paths containing URLs.",
        )
        self.parser.add_argument(
            "--subs",
            type=str,
            default="",
            help='Comma-separated list of subreddits (e.g., "r/python,r/askreddit").',
        )
        self.parser.add_argument(
            "--multis",
            type=str,
            default="",
            help='Comma-separated list of multireddits (e.g., "m/programming").',
        )

        self.args = self.parser.parse_args()

        self.urls: List[str] = _parse_csv(self.args.urls)
        self.src_files: List[str] = _parse_csv(self.args.src_files)
        self.subs: List[str] = _parse_csv(self.args.subs)
        self.multis: List[str] = _parse_csv(self.args.multis)

        # You can add some simple logging or validation below if needed:
        logger.info("Parsed %d URL(s) from --urls", len(self.urls))
        logger.info("Parsed %d file(s) from --src-files", len(self.src_files))
        logger.info("Parsed %d subreddit(s) from --subs", len(self.subs))
        logger.info("Parsed %d multireddit(s) from --multis", len(self.multis))
